### PR TITLE
Improve series creation and fetching 

### PR DESCRIPTION
**Pull request by cevian**
This commit changes the method we use to fetch and create series.
Before we created series by passing down the array of names and
values directly. This necessited for us to create a series in
it's own transaction because otherwise we could get a deadlock
between label creation ordering of different series.

Now we switched to doing this in 2 parts:
1) create all the labels (sorted beforehand)
2) create series by passing down the label arrays constructed
   from the label ids fetched in step 1.

This has two advantages:
1) in micro-benchmarks this reduced series creation/fetching time by 3x
2) it significantly reduces the number of txns

The one disadvantage is that this necessitates 2 round-trips,
to the db per batch instead of one. C'est la vie.

### CODE REVIEW

File: pkg/pgmodel/ingestor/handler.go [hunk 1]
```diff
@@ -72,113 +76,177 @@ func (h *insertHandler) flushPending() {
...
 		return fmt.Errorf("Error setting series ids: %w", err)
 	}
-	defer br.Close()
+	if len(labelArraySet) == 0 {
+		return nil
+	}
 
-	// BEGIN;
-	_, err = br.Exec()
+	labelArrayArray := pgtype.NewArrayType("prom_api.label_array[]", h.labelArrayOID, func() pgtype.ValueTranscoder { return &pgtype.Int4Array{} })
```
Comment (blagojts) [inline pkg/pgmodel/ingestor/handler.go:143]: can we create this once per insert handler?

**Review by blagojts — APPROVED** [review]

**Comment by cevian** [comment]
Some performance results. 
On an overfilled cache (thus fetching mostly already created series): 
insert rate with this change: 40-30k
insert rate before this change: 18k at best

to fill the cache with changes; 2 min and a bit
before change 3 min and a bit

File: pkg/migrations/sql/idempotent/base.sql [hunk 2]
```diff
@@ -740,6 +740,33 @@ COMMENT ON FUNCTION SCHEMA_CATALOG.get_or_create_label_array(text, text[], text[
...
+            LEFT JOIN SCHEMA_CATALOG.label_key_position lkp
+               ON
+               (
+                  lkp.metric_name = get_or_create_label_ids.metric_name AND
+                  lkp.key = kv.key
+               )
+$$
+LANGUAGE SQL VOLATILE;
+COMMENT ON FUNCTION SCHEMA_CATALOG.get_or_create_label_ids(text, text[], text[])
+IS 'converts a metric name, array of keys, and array of values label ids';
```
Comment (Harkishen-Singh) [inline pkg/migrations/sql/idempotent/base.sql:766]: I think the comment isn't explanatory. Converts to what?

File: pkg/migrations/sql/idempotent/base.sql [hunk 3]
```diff
@@ -740,6 +740,33 @@ COMMENT ON FUNCTION SCHEMA_CATALOG.get_or_create_label_array(text, text[], text[
...
 
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.get_or_create_label_ids(metric_name TEXT, label_keys text[], label_values text[])
+RETURNS TABLE(pos int, id int, label_key text, label_value text) AS $$
+        SELECT
+            -- only call the functions to create new key positions
+            -- and label ids if they don't exist (for performance reasons)
+            coalesce(lkp.pos,
+              SCHEMA_CATALOG.get_or_create_label_key_pos(get_or_create_label_ids.metric_name, kv.key)) idx,
+            coalesce(l.id,
+              SCHEMA_CATALOG.get_or_create_label_id(kv.key, kv.value)) val,
```
Comment (Harkishen-Singh) [inline pkg/migrations/sql/idempotent/base.sql:751]: Haven't gone into details, but is 749 and 751 dependent on each other? (if yes, then I hope they are in proper sync).

File: pkg/pgmodel/ingestor/handler.go [hunk 4]
```diff
@@ -72,113 +76,179 @@ func (h *insertHandler) flushPending() {
...
+				if !ok {
+					labelMap[key] = labelInfo{}
+					labelList.Add(names[i], values[i])
+				}
+			}
 		}
-
-		batch.Queue("BEGIN;")
-		batch.Queue(getSeriesIDForLabelSQL, curr.GetSeries().MetricName(), names, values)
-		batch.Queue("COMMIT;")
```
Comment (Harkishen-Singh) [inline pkg/pgmodel/ingestor/handler.go:121]: In my cpu profilings, i did get issues here. Removing these should be very helpful.

File: pkg/pgmodel/ingestor/handler.go [hunk 5]
```diff
@@ -72,113 +76,179 @@ func (h *insertHandler) flushPending() {
...
-			//was already set
-			continue
-		}
-		if lastSeenLabel != nil && lastSeenLabel.Equal(curr.GetSeries()) {
-			batchSeries[len(batchSeries)-1] = append(batchSeries[len(batchSeries)-1], curr)
-			continue
+	metricName := seriesToInsert[0].MetricName()
+	labelMap := make(map[labels.Label]labelInfo, len(seriesToInsert))
+	labelList := model.NewLabelList(len(seriesToInsert))
+	{
```
Comment (Harkishen-Singh) [inline pkg/pgmodel/ingestor/handler.go:104]: Not sure but is this required? (We were against this in prom-migrator)

File: pkg/pgmodel/ingestor/handler.go [hunk 6]
```diff
@@ -72,113 +76,179 @@ func (h *insertHandler) flushPending() {
...
+	}
+	res, err := h.conn.Query(context.Background(), "SELECT r.series_id, l.nr FROM unnest($2::prom_api.label_array[]) WITH ORDINALITY l(elem, nr) INNER JOIN LATERAL _prom_catalog.get_or_create_series_id_for_label_array($1, l.elem) r ON (TRUE)", metricName, labelArrayArray)
+	if err != nil {
+		return fmt.Errorf("Error setting series_id: cannot query for series_id: %w", err)
+	}
+	defer res.Close()
+
+	for res.Next() {
+		var id model.SeriesID
+		var ordinality int64
```
Comment (Harkishen-Singh) [inline pkg/pgmodel/ingestor/handler.go:158]: we should merge into single var block.

File: pkg/pgmodel/ingestor/insert.go [hunk 7]
```diff
@@ -60,7 +60,7 @@ func initializeInserterRoutine(conn pgxconn.PgxConn, metricName string, complete
 	return tableName, err
 }
 
-func runInserterRoutine(conn pgxconn.PgxConn, input chan *insertDataRequest, metricName string, completeMetricCreationSignal chan struct{}, metricTableNames cache.MetricCache, toCopiers chan copyRequest) {
+func runInserterRoutine(conn pgxconn.PgxConn, input chan *insertDataRequest, metricName string, completeMetricCreationSignal chan struct{}, metricTableNames cache.MetricCache, toCopiers chan copyRequest, labelArrayOID uint32) {
```
Comment (Harkishen-Singh) [inline pkg/pgmodel/ingestor/insert.go:63]: (optional) This line is already too long. I think we should switch to a struct for this.

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
DIdn't go much deep, but should be good to go except for a few nits.

File: pkg/migrations/sql/idempotent/base.sql (hunk 3 above)
Comment (cevian) [reply pkg/migrations/sql/idempotent/base.sql:751]: There should be no dependency here

File: pkg/pgmodel/ingestor/handler.go (hunk 5 above)
Comment (cevian) [reply pkg/pgmodel/ingestor/handler.go:104]: This is inline, purely for performance reasons. added a comment

File: pkg/pgmodel/ingestor/insert.go (hunk 7 above)
Comment (cevian) [reply pkg/pgmodel/ingestor/insert.go:63]: fixed formatting; leaving change to struct later

**Review by Harkishen-Singh — APPROVED** [review]

**Pull request merged**